provider/cmd/crd2pulumi/schema.go

```diff
@@ -0,0 +1,303 @@
+// Union type of integer and string
+var IntOrStringTypeSpec = pschema.TypeSpec{
+	OneOf: []pschema.TypeSpec{
+		pschema.TypeSpec{
+			Type: "integer",
+		},
+		pschema.TypeSpec{
+			Type: "string",
+		},
+	},
+}
+
+// Returns true if the given TypeSpec is of type any; returns false otherwise
+func isAnyType(typeSpec pschema.TypeSpec) bool {
+	return typeSpec.Ref == AnyTypeRef
+}
+
+// GetObjectTypeSpecs generates types for each versioned schema. Returns a
+// mapping from each type's name in the format of
+// "{.spec.group}/{.spec.versions[*].name}:{.spec.names.kind}" to its proper
+// pschema.ObjectTypeSpec.
+func GetObjectTypeSpecs(versions map[string]map[string]interface{}, name, kind string) map[string]pschema.ObjectTypeSpec {
+	objectTypeSpecs := map[string]pschema.ObjectTypeSpec{}
+	for version, schema := range versions {
+		baseRef := fmt.Sprintf("crds:%s/%s:%s", name, version, kind)
+		addType(schema, baseRef, objectTypeSpecs)
+		// Adds "Args" to the baseRef name
+		newBaseRef := baseRef + "Args"
+		objectTypeSpecs[newBaseRef] = objectTypeSpecs[baseRef]
+		objectTypeSpecs[newBaseRef].Properties["metadata"] = pschema.PropertySpec{
+			TypeSpec: pschema.TypeSpec{
+				Ref: ObjectMetaRef,
+			},
+		}
+		delete(objectTypeSpecs, baseRef)
+	}
+
+	return objectTypeSpecs
+}
+
+// addType converts the given OpenAPI `schema` to a ObjectTypeSpec and adds it
+// to the `types` map under the given `name`. Recursively converts and adds all
+// nested schemas as well.
+func addType(schema map[string]interface{}, name string, types map[string]pschema.ObjectTypeSpec) {
+	properties, foundProperties, _ := unstruct.NestedMap(schema, "properties")
+	if !foundProperties {
+		return
+	}
+
+	description, _, _ := unstruct.NestedString(schema, "description")
+	schemaType, _, _ := unstruct.NestedString(schema, "type")
+	required, _, _ := unstruct.NestedStringSlice(schema, "required")
+
+	propertySpecs := map[string]pschema.PropertySpec{}
+	for propertyName := range properties {
+		propertySchema, _, _ := unstruct.NestedMap(properties, propertyName)
+		propertyDescription, _, _ := unstruct.NestedString(propertySchema, "description")
+		defaultValue, _, _ := unstruct.NestedFieldNoCopy(propertySchema, "default")
+		propertySpecs[propertyName] = pschema.PropertySpec{
+			TypeSpec:    getTypeSpec(propertySchema, name+strings.Title(propertyName), types),
+			Description: propertyDescription,
+			Default:     defaultValue,
+		}
+	}
+
+	types[name] = pschema.ObjectTypeSpec{
+		Type:        schemaType,
+		Properties:  propertySpecs,
+		Required:    required,
+		Description: description,
+	}
+}
+
+// Replaces the hyphens in the x-kubernetes-... fields with underscores
+func underscoreFields(schema map[string]interface{}) {
+	for field, val := range schema {
+		for _, hyphenedField := range HyphenedFields {
+			if field == hyphenedField {
+				delete(schema, field)
+				underScoredField := strings.ReplaceAll(field, "-", "_")
+				schema[underScoredField] = val
+			}
+		}
```

Comment:
It'd be slightly more efficient to make `HyphenedFields` a map, e.g.:

```go
var HyphenedFields = map[string]struct{}{
	"x-kubernetes-embedded-resource":       struct{}{},
	"x-kubernetes-int-or-string":           struct{}{},
	"x-kubernetes-list-map-keys":           struct{}{},
	"x-kubernetes-list-type":               struct{}{},
	"x-kubernetes-map-type":                struct{}{},
	"x-kubernetes-preserve-unknown-fields": struct{}{},
}
```

And lookup:

```go
		if _, ok := HyphenedFields[field]; ok {
			delete(schema, field)
			underScoredField := strings.ReplaceAll(field, "-", "_")
			schema[underScoredField] = val
		}
```

Reply:
Even easier is using `codegen.StringSet` from pulumi/pulumi. It uses the `map[string]struct{}` construction under the hood, but the interface is clearer.